dev/PushNotifications/PushManager.cpp

```diff
@@ -4,11 +4,111 @@
 #include "pch.h"
 #include "PushManager.h"
 #include "PushManager.g.cpp"
+#include <winrt\windows.networking.pushnotifications.h>
+#include <winerror.h>
+#include <algorithm>
+#include <list>
+#include <iostream>
+#include <ChannelResult.h>
+#include <winerror.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+// MAX BACKOFF TIME IS 16MINUTES 
+#define MAX_BACKOFF_SECONDS 960
+
+// Error code - platform is not connected
+#define WNP_E_NOT_CONNECTED 0x880403E8L
+
+// Error code - platform reconnecting with wns
+#define WNP_E_RECONNECTING  0x880403E9L
+
+// Error code - Current bind operation failed because another bind operation in progress
+#define WNP_E_BIND_USER_BUSY 0x880403FEL
 
 namespace winrt::Microsoft::ProjectReunion::implementation
 {
+    std::list<winrt::guid> PushManager::remoteIdList;
+    std::mutex PushManager::mtx;
+    std::unique_lock<std::mutex> PushManager::lock (mtx, std::defer_lock); // Avoid locking during the constructor call
+
+    bool PushManager::isChannelRequestRetryable(const winrt::hresult& hrException)
+    {
+        switch (hrException)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+            // E_TIMEOUT
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING
+        case WNP_E_BIND_USER_BUSY:
+            return true;
+        default:
+            return false;
+        }
+    }
+
     Windows::Foundation::IAsyncOperationWithProgress<Microsoft::ProjectReunion::ChannelResult, Microsoft::ProjectReunion::ChannelResult> PushManager::CreateChannelAsync(winrt::guid remoteId)
```

Comment:
Can we verify the implications of lifetime management? What would happen if the process exited but this AsyncProgress operation is not yet complete? Is the process termination graceful?

Reply:
https://en.cppreference.com/w/cpp/language/coroutines - When the coroutine state is destroyed either because it terminated via co_return or uncaught exception, or because it was destroyed via its handle, it does the following:

calls the destructor of the promise object.
calls the destructors of the function parameter copies.
calls operator delete to free the memory used by the coroutine state
transfers execution back to the caller/resumer.

So the coroutines completes gracefully.

Also it is important that the caller of the coroutine API: CreateChannelAsync will have to call IAsyncOperationWithProgress.Cancel to stop the asynchronous operation before the caller process exits. We would be clearly mentioning this in the sample that we are providing to cancel the asyncoperation if you are going out of scope. Thanks.